proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+We pass the desired data constructor as a type to `reifyInfoTablePtr#`, using type ascription for the return value:
+
+```haskell
+let leftInfoPtr :: InfoTablePtrOf# 'Left = reifyInfoTablePtr# (# #)
+```
+
+`reifyInfoTablePtr#` is a compile-time primops (i.e. substituted with actual Cmm code in the _Stg to Cmm_ phase). It inspects the type of its return value (which is accessible in the code generation pipeline for compile-time primops `emitPrimOp`):
+
+- if it is of shape `InfoTablePtrOf# 〈s〉`, where `〈s〉` is a `Symbol`, it returns the info table pointer for the Stg object `stg_〈s〉`
+- if it is of shape `InfoTablePtrOf# '〈d〉`, where `〈d〉` is a data constructor, it returns the info table pointer for data constructor `〈d〉`
+- otherwise, it raises a compile-time error (I've yet to find how to do that in practice)
+
+Defining `reifyInfoTablePtr#` requires support for kind-polymorphic type variables in primop generation. I include the following changes in the proposal, following what has been done for levity-polymorphic type variables:
+
+```haskell
+kind1TyVar, kind2TyVar :: TyVar
+(kind1TyVar : kind2TyVar : _)
+  = mkTemplateKindVars (repeat liftedTypeKind)
+
+kind1TyVarInf, kind2TyVarInf :: TyVarBinder
+kind1TyVarInf = mkTyVarBinder Inferred kind1TyVar
+kind2TyVarInf = mkTyVarBinder Inferred kind2TyVar
+
+kind1Ty, kind2Ty :: Type
+kind1Ty = mkTyVarTy kind1TyVar
+kind2Ty = mkTyVarTy kind2TyVar
+
+kindPolyAlphaTyVar, kindPolyBetaTyVar :: TyVar
+[kindPolyAlphaTyVar, kindPolyBetaTyVar] =
+  mkTemplateTyVars
+    [ kind1Ty
+    , kind2Ty]
+
+kindPolyAlphaTyVarSpec, kindPolyBetaTyVarSpec :: TyVarBinder
+kindPolyAlphaTyVarSpec = mkTyVarBinder Specified kindPolyAlphaTyVar
+kindPolyBetaTyVarSpec  = mkTyVarBinder Specified kindPolyBetaTyVar
+
+kindPolyAlphaTy, kindPolyBetaTy :: Type
+kindPolyAlphaTy = mkTyVarTy kindPolyAlphaTyVar
+kindPolyBetaTy  = mkTyVarTy kindPolyBetaTyVar
+```
+
+### Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family
+
+```haskell
+type family LiftedCtorToSymbol (liftedCtor :: k) :: Symbol
+```
+
+When making an API around `compactAddHollow#`, we need to know the number and type of the fields that the constructor has. That information can be easily retrieved with `GHC.Generics`. However, the type-level structure returned by `GHC.Generics` distinguishes constructors of a given type by their name, as a `Symbol`.
+
+I introduce a built-in type family `LiftedCtorToSymbol` whose in charge of translating a data constructor lifted as a type into the corresponding symbol. That way a library can leverage `GHC.Generics` to allow direct and safe structure building in compact regions for most user-defined data types.
+
+For example:
+
+```haskell
+LiftedCtorToSymbol 'Left ~ "Left"
+```
+
+More generally, forall data constructor `〈d〉`,
+
+```haskell
+LiftedCtorToSpecCtor '〈d〉 ~ "〈d〉"
+```
```

Comment:
Honestly, this example isn't clear (it's not even really an example). Just spell out the semantics in English, using the GHC terminology (occurrence name and original name, as above).

Reply:
Removed it.